lib/debug/ProfilingPlugin.js

```diff
@@ -0,0 +1,394 @@
+			this.sendCommand("Profiler.start"),
+		]);
+	}
+
+	sendCommand(method, params) {
+		if(this.hasSession()) {
+			return new Promise((res, rej) => {
+				return this.session.post(method, params, (err, params) => {
+					if(err !== null) {
+						rej(err);
+					} else {
+						res(params);
+					}
+				});
+			});
+		} else {
+			return Promise.resolve();
+		}
+	}
+
+	destroy() {
+		if(this.hasSession()) {
+			this.session.disconnect();
+		}
+
+		return Promise.resolve();
+	}
+
+	stopProfiling() {
+		return this.sendCommand("Profiler.stop");
+	}
+}
+
+function sleep(num) {
+	return new Promise((res) => {
+		setTimeout(res, num);
+	});
+}
+
+/**
+ * @param {string} outPath The location where to write the log.
+ * @returns {{trace: ?,	counter: number, profiler: Profiler}} The trace object
+ */
+function createTrace(outPath) {
+	const trace = new Trace({
+		noStream: true
+	});
+	const profiler = new Profiler(inspector);
+
+	let counter = 0;
+
+	trace.pipe(fs.createWriteStream(outPath));
+	// These are critical events that need to be inserted so that tools like
+	// chrome dev tools can load the profile.
+	trace.instantEvent({
+		name: "TracingStartedInPage",
+		id: ++counter,
+		cat: ["disabled-by-default-devtools.timeline"],
+		args: {
+			data: {
+				sessionId: "-1",
+				page: "0xfff",
+				frames: [{
+					frame: "0xfff",
+					url: "webpack",
+					name: ""
+				}]
+			}
+		}
+	});
+
+	trace.instantEvent({
+		name: "TracingStartedInBrowser",
+		id: ++counter,
+		cat: ["disabled-by-default-devtools.timeline"],
+		args: {
+			data: {
+				sessionId: "-1"
+			},
+		}
+	});
+
+	return {
+		trace,
+		counter,
+		profiler
+	};
+}
+
+class ProfilingPlugin {
+	// TODO: Add plugin schema validation here since there are options.
+	constructor(opts) {
+		opts = opts || {};
+		this.outPath = opts.outPath || "events.json";
+	}
+
+	apply(compiler) {
+		const tracer = createTrace(this.outPath);
+		tracer.profiler.startProfiling();
+
+		// Compiler Hooks
+		Object.keys(compiler.hooks).forEach(hookName => {
+			compiler.hooks[hookName].intercept(makeInterceptorFor("Compiler", tracer)(hookName));
+		});
+
+		Object.keys(compiler.resolverFactory.hooks).forEach(hookName => {
+			compiler.resolverFactory.hooks[hookName].intercept(makeInterceptorFor("Resolver", tracer)(hookName));
+		});
+
+		compiler.hooks.compilation.tap("ProfilingPlugin", (compilation, {
+			normalModuleFactory,
+			contextModuleFactory
+		}) => {
+			interceptAllHooksFor(compilation, tracer, "Compilation");
+			interceptAllHooksFor(normalModuleFactory, tracer, "Normal Module Factory");
+			interceptAllHooksFor(contextModuleFactory, tracer, "Context Module Factory");
+			interceptAllParserHooks(normalModuleFactory, tracer);
+			interceptTemplateInstancesFrom(compilation, tracer);
+		});
+
+		// We need to write out the CPU profile when we are all done.
+		compiler.hooks.done.tap("Profiling plugin", () => {
+			// Wait just in case someone else registered an after done hook.
```

Comment:
``` js
compiler.hooks.done.tap({
  name: "Profiling plugin",
  stage: Infinity
}, () => {
```

This registers itself as the last plugin.

Reply:
neat, thanks for the tip.

Reply:
Ahhhhh _thats_ how we set it.